[OpenMP] [amdgpu] Added a synchronous version of data exchange.

dhruvachak opened the pull request:
Similar to H2D and D2H, use synchronous mode for large data transfers
beyond a certain size for D2D as well. As with H2D and D2H, this size is
controlled by an env-var.

dhruvachak committed:
[OpenMP] [amdgpu] Added a synchronous version of data exchange.

Similar to H2D and D2H, use synchronous mode for large data transfers
beyond a certain size for D2D as well. As with H2D and D2H, this size is
controlled by an env-var.

jhuber6 commented:
Can you update the description to describe why we want synchronous data exchange?

dhruvachak commented:
I amended the commit message to include the following:

Similar to H2D and D2H, use synchronous mode for large data transfers
beyond a certain size for D2D as well. As with H2D and D2H, this size is
controlled by an env-var.

arsenm commented on openmp/libomptarget/test/offloading/d2d_memcpy_sync.c:
```diff
@@ -0,0 +1,67 @@
+  int *src_ptr = omp_target_alloc(length, src_device);
+  int *dst_ptr = omp_target_alloc(length, dst_device);
+
+  assert(src_ptr && "src_ptr is NULL");
+  assert(dst_ptr && "dst_ptr is NULL");
```
don't assert to check allocation failure

dhruvachak replied:
Changed to check and FAIL if required. The earlier change was based on an existing test which asserted.

arsenm commented on openmp/libomptarget/test/offloading/d2d_memcpy_sync.c:
```diff
@@ -0,0 +1,67 @@
+
+  int *buffer = malloc(length);
+
+  assert(buffer && "failed to allocate host buffer");
```
same

dhruvachak replied:
done.

dhruvachak committed:
[OpenMP] [amdgpu] Added a synchronous version of data exchange.

Changed test to not assert on allocation failure. Instead it checks for
that condition and returns a failure status.

jhuber6 commented:
> I amended the commit message to include the following:
> 
> Similar to H2D and D2H, use synchronous mode for large data transfers beyond a certain size for D2D as well. As with H2D and D2H, this size is controlled by an env-var.

You should do it to the PR message as well. When the PR is merged it will squash the commits and use the PR's description instead.

dhruvachak commented:
> > I amended the commit message to include the following:
> > Similar to H2D and D2H, use synchronous mode for large data transfers beyond a certain size for D2D as well. As with H2D and D2H, this size is controlled by an env-var.
> 
> You should do it to the PR message as well. When the PR is merged it will squash the commits and use the PR's description instead.

Yes, I just realized that the PR comment was not updated. Now it should be updated.

jhuber6 reviewed (approved):
LG, thanks

What's the expected behavior when doing a D2D memcpy onto the same device? Just wondering if that's a no-op under the hood for the purposes of the buildbot.

dhruvachak commented:
> LG, thanks
> 
> What's the expected behavior when doing a D2D memcpy onto the same device? Just wondering if that's a no-op under the hood for the purposes of the buildbot.

The plugin still uses the same async memcpy. I don't know whether there is some underlying optimization for the same-device case.
